### ISSUE TITLE
 Specification pattern issue

### ISSUE THREAD

**Issue by mrs01dev** [SteveDunn/Vogen#272]
### Describe the bug

This great library works Microsoft EntityFrameworkCore successfully in  database   query.

but when using a package "Ardalis.Specification"  to query database    , Microsoft EntityFrameworkCore  result in Cast issues  like  [
Invalid cast from 'System.Int64'
System.InvalidCastException
]

### Steps to reproduce

add this package to Database Query  project
https://github.com/ardalis/specification
Help :
http://specification.ardalis.com/usage/create-specifications.html
Query SQL server Database

### Expected behaviour

return an object from data base  that have field consist of "ValueObject" with  type long

**Comment by SteveDunn** [comment]
Hi @mrs01dev - thanks for the bug report. Could you share your example project where you are having problems?

I took a quick look at the project you mentioned and changed `Customer::Id` from an `int` to a `CustomerId`, and it compiled OK (after changing the `LangVersion` to `10`), and the tests seem to run OK.

Did you decorate your Value Object with the EF Core conversion? e.g.

```csharp
  [ValueObject(conversions: Conversions.EfCoreValueConverter)]
  public partial struct CustomerId
  {
  }
```

**Comment by jeffward01** [comment]
> but we need to reference EFCore, which is `net6.0`. This is one of the reasons that I don't tend to use Value Objects outside of the domain and use primitives at the infrastructure layer.

Hello!

Just wanted to ask for some clarification on this, so technically speaking - lets say I have these projects:

* MyBookStore.API
* MyBookStore.Core
* MyBookStore.Infra
   * This is where `EfCore` resides, and my DbContext
* MyBookStore.Domain
   * Has Vogen installed, no other proejcts have Vogen installed
* MyBookStore.SharedCommon

The project (assembly) `MyBookStore.Infra` will have a reference and depend on the `MyBookStore.Domain`, which only has (1) dependency of Vogen, and nothing else.


**Questions**
**1.) Scenario One** EfCore directly installed in same project that uses Vogen

- In this example, the only project which utilizes `Vogen`, and has `Vogen` installed is `*.Domain`

**Question One**
Do you mean that for Vogen to generate `EfCoreValueConverter` correctly, the assembly `MyBookStore.Domain` needs to also have a reference to EfCore?  

**2.) Scenario Two** EfCore is installed in a separate project, which is referenced by the project that has Vogen installed

Let's say that, the only project that has `EfCore installed, is `*.SharedCommon`

* `*.Domain` project had a reference to `*.SharedCommon`.
* `*.Infra` project has a reference to `*.SharedCommon` and `*.Domain`

**Question Two**
Is this acceptable, or does the source generate require that the direct assembly `*.Domain` has a reference to `EFCore` in its own assembly (not from a referenced project)?
 

---

I think the above example I made is confusing, sorry. 

Basically, does Vogen require the project that has Vogen installed to have EfCore installed _directly_ as well, or can the project consuming (has Vogen installed) Vogen, reference a _different_ project which has EfCore installed and that is acceptable?

**Comment by Blackclaws** [comment]
@jeffward01 The assembly that defines the converters needs to have a reference to EfCore. This is because the converters inherit Microsoft.EntityFrameworkCore.Storage.ValueConversion.ValueConverter.

So yes, you need a reference to EfCore in the assembly that defines the Vogen types that you want to generate value converters for.

**Comment by Blackclaws** [comment]
> Hi @mrs01dev - thanks for the bug report. Could you share your example project where you are having problems?
> 
> I took a quick look at the project you mentioned and changed `Customer::Id` from an `int` to a `CustomerId`, and it compiled OK (after changing the `LangVersion` to `10`). I couldn't run the tests though as the types were in `Ardalis.SampleApp.Core` which targets `netstandard2.0`, but we need to reference EFCore, which is `net6.0`. This is one of the reasons that I don't tend to use Value Objects outside of the domain and use primitives at the infrastructure layer.
> 
> But anyway, I think if you decorate your Value Objects with the EF Core it'll work - e.g.
> 
> ```cs
>   [ValueObject(conversions: Conversions.EfCoreValueConverter)]
>   public partial struct CustomerId
>   {
>   }
> ```

@SteveDunn I get why you'd want to keep your domain as clean as possible, however if you have to map your domain types once more to infrastructure types (which I'm well aware is done to keep the separation as wide as possible) you end up with one more layer of maintenance and glue code that is sometimes just a 1:1 correspondence. Also since with ValueConverters you end up with a rather easy way to actually use these strongly typed primitives in EfCore the benefits of directly using them outweigh the drawbacks I think.

That being said, having to register each value converter separately is a bit annoying especially since its just boilerplate. I'd love for there to be a central static function that I can just pass the model builder into that handles all the registrations for all Value Types in an assembly.

An alternative to the problem of having to add EfCore as a dependency on the Domain level would be to expose the Deserialize method as public and then have a new attribute that creates value converters all at once in a class in the Infrastructure assembly (which could also contain a central method to register _all_ of them).

**Comment by jeffward01** [comment]
@Blackclaws I have a central “register all function” I can share with you later today. I’m not at my computer. 

Strangely, I believe I am able to generate EFCore converters without the reference to the EFCore library. I’m not at the computer, I was this morning, and I’ll have to confirm it. 

One sorta “gotcha” regarding EFCore and vogen is that Vogen will throw an error if the object is initialized with no value, or rather, if it’s not initialized at all. 

So for example, I create a new “book” and I attach it to the author entity (without setting the foriegn bookID property on the author). I do this all the time, becuse EFCore is smart (hahah it tries to be) And will set the fk for me. 

If I do that with vogen, I’ll get an exception because the value is not initialized 

I have generic value generators I’ve written for vogen as well, I have not tested them deeply, but I was having issues with getting them to generate values in the above scenario. The value generators work fine for primary keys. 

I can share those as well. Give me a few hours!

Also, i would love someone’s help to test and see if the value generator works on normal properties. I used UUID7 for example, so in the author book Scenario, the value generator was not working. If someone could test this (or I can if I have time) that would be great. I’m just a bit tight on time this week. 

If I get to testing it I for sure will

**Comment by jeffward01** [comment]
Another annoying issue with value objects is querying using a repository or the DbContext. 

I haven’t solved it yet (I just discovered the issue this AM). 

Basically, you can’t access the .Value property using EFCore, and EFCore to my knowledge won’t assign a value to your vogen type until you query executes.. 

So what I plan on doing is using one of those expression libraries, to execute my query expression without executing the query. 

I forget the name of the library, I think it’s called linq extensions, or EFCore extensions. Something like that. 

This is another gotcha, it has nothing to do with Vogen, it’s EFCore. It’s the same reason why EFCore really sucks with Maybe<> types. 

Another solution is a private readonly value type property, then you access it using manually (yuck) using EF.Property(“BookId”) or the property name, if it’s a private field, just use “_bookId”. So
I’m an advocate for the repository pattern ah ha! And I can put a helper method to translate ids and call EF.property globally within a repository method (take that repo haters). So that works… not the best solution but definitely works  


So those are two gotchas I have discovered when working with EFCore and Vogen. 

If someone has better solutions please let me (us) know

**Comment by Blackclaws** [comment]
> Another annoying issue with value objects is querying using a repository or the DbContext.
> 
> I haven’t solved it yet (I just discovered the issue this AM).
> 
> Basically, you can’t access the .Value property using EFCore, and EFCore to my knowledge won’t assign a value to your vogen type until you query executes..
> 
> So what I plan on doing is using one of those expression libraries, to execute my query expression without executing the query.
> 
> I forget the name of the library, I think it’s called linq extensions, or EFCore extensions. Something like that.
> 
> This is another gotcha, it has nothing to do with Vogen, it’s EFCore. It’s the same reason why EFCore really sucks with Maybe<> types.
> 
> Another solution is a private readonly value type property, then you access it using manually (yuck) using EF.Property(“BookId”) or the property name, if it’s a private field, just use “_bookId”. So I’m an advocate for the repository pattern ah ha! And I can put a helper method to translate ids and call EF.property globally within a repository method (take that repo haters). So that works… not the best solution but definitely works
> 
> So those are two gotchas I have discovered when working with EFCore and Vogen.
> 
> If someone has better solutions please let me (us) know

So I personally use Ardalis.Specification and the Repository from that library that accepts general queries. You can use a Where clause to compare value objects with registered value converters and the clause will be executed in the database. You don't even need to access .Value to compare them. You can do the same by just calling .Where on a DbSet.

You get queries like this:
```
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT COUNT(*)
      FROM "Entities" AS "e"
      WHERE "e"."ValueObject" = 'Wow'
```

**Comment by jeffward01** [comment]
> > Another annoying issue with value objects is querying using a repository or the DbContext.
> > I haven’t solved it yet (I just discovered the issue this AM).
> > Basically, you can’t access the .Value property using EFCore, and EFCore to my knowledge won’t assign a value to your vogen type until you query executes..
> > So what I plan on doing is using one of those expression libraries, to execute my query expression without executing the query.
> > I forget the name of the library, I think it’s called linq extensions, or EFCore extensions. Something like that.
> > This is another gotcha, it has nothing to do with Vogen, it’s EFCore. It’s the same reason why EFCore really sucks with Maybe<> types.
> > Another solution is a private readonly value type property, then you access it using manually (yuck) using EF.Property(“BookId”) or the property name, if it’s a private field, just use “_bookId”. So I’m an advocate for the repository pattern ah ha! And I can put a helper method to translate ids and call EF.property globally within a repository method (take that repo haters). So that works… not the best solution but definitely works
> > So those are two gotchas I have discovered when working with EFCore and Vogen.
> > If someone has better solutions please let me (us) know
> 
> So I personally use Ardalis.Specification and the Repository from that library that accepts general queries. You can use a Where clause to compare value objects with registered value converters and the clause will be executed in the database. You don't even need to access .Value to compare them. You can do the same by just calling .Where on a DbSet.
> 
> You get queries like this:
> 
> ```
>       Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
>       SELECT COUNT(*)
>       FROM "Entities" AS "e"
>       WHERE "e"."ValueObject" = 'Wow'
> ```

Im pretty sure you'll still have issues regarding those value objects and EfCore,  Its a LINQ / IQueryable limitation where the `.Value` will not be examined, because it has to be compiled.  I used a library called to solve the issues. https://github.com/scottksmith95/LINQKit

This is how I use LinqKit:

```csharp
        Guid id = Guid.Parse("0185C53F-02F5-714B-B293-CC826DBFADB5");
        WeatherForecastId weatherId = WeatherForecastId.From(id);
   Maybe<WeatherForecast> weatherForecast3 = await this._unitOfWork.GetRepositoryAsync<WeatherForecast>()
                .FirstOrDefaultAsync(_ => _.Id == weatherId, cancellationToken)
                .ConfigureAwait(false);
```


```csharp
    /// <inheritdoc />
    public async Task<Maybe<T>> FirstOrDefaultAsync(
        Expression<Func<T, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        IEnumerable<T> enumerable = this._dbSet.AsQueryable()
            .WithQuery(predicate);

        return await Task.FromResult(enumerable.TryFirst())
            .ConfigureAwait(false);
    }

    public static IEnumerable<T> WithQuery<T>(
        this IQueryable<T> query,
        Expression<Func<T, bool>> predicate)
        where T : class
    {
        ExpressionStarter<T>? linqKitPredicate = PredicateBuilder.New<T>();

        linqKitPredicate.And(predicate);

        return query.AsEnumerable()
            .Where(linqKitPredicate.Compile());
    }
```

And that solved the issue for me. Feel free to give it a go and see if it works ok for you.  I really like Ardalis.Specification, also his Endpoints is one of my favorite, and I use his Guards daily.

I ended up re-writing the Vogen EfCore Converter for `Guid` as `GUID` is what I use for my primary keys.  You'll have to modify the code or extend it so that it works with `int`.  My code was just a quick and simple implementation for myself, so I didn't do anything fancy with it.

I left the debugging comments in it as well, feel free to remove them.

This applies a Vogen Value Converter and Comparer:

```csharp

public static class VogenValueObjectModelBuilderExtensions
{
    public static void UseVogenStronglyTypedId(
        this EntityTypeBuilder entityTypeBuilder,
        bool hasNewInitializer)
    {
        Console.WriteLine("#######################################################################################");
        Console.WriteLine("#######################################################################################");
        Console.WriteLine(
            $"############  Configuring Entity >>>   {entityTypeBuilder.Metadata.ClrType.Name} ####################");
        Console.WriteLine("#######################################################################################");
        Console.WriteLine("#######################################################################################");

        IEnumerable<PropertyInfo> properties = entityTypeBuilder.Metadata.ClrType.GetProperties()
            .Where(
                propertyInfo => propertyInfo.PropertyType.UnwrapNullableType()
                    .IsVogenValueObject());
 

        foreach (PropertyInfo property in properties)
        {
            if (property.IsVogenValueObject()
                .IsFalse())
            {
                continue;
            }
            
            Type idType = property.PropertyType;
            Type valueType = idType.GetVogenValueObjectType();
            Console.WriteLine(
                ">>>> Property Name {2} ||||Primary Type: {0} ||| Vogen Value (underlying) Type: {1}",
                idType.Name,
                valueType.Name,
                property.Name);

            _addValueGenerator(idType, valueType, property);
            _addValueConverter(idType, valueType, property);
 
        }

        void _addValueGenerator( Type vogenObjectType,
            Type valueType,
            PropertyInfo property)
        {
            if (valueType != typeof(Guid))
            {
                return;
            }
            
            
            Type converterTypeTemplate = typeof(VogenDatabaseValueGenerator<>);
            Type valueGeneratorType = converterTypeTemplate.MakeGenericType(vogenObjectType);
            ValueGenerator? converter = (ValueGenerator)Activator.CreateInstance(valueGeneratorType);

 
            entityTypeBuilder.Property(property.Name)
                .ValueGeneratedOnAdd()
                .HasValueGenerator(converter.GetType());
      

        }
        
        void _addValueConverter(
            Type vogenObjectType,
            Type valueType,
            PropertyInfo property)
        {
     
            Type converterTypeTemplate = typeof(VogenStronglyTypedIdConverter<,>);

            Type converterType = converterTypeTemplate.MakeGenericType(vogenObjectType, valueType);
            ValueConverter? converter = (ValueConverter)Activator.CreateInstance(converterType, (object)null);

            Console.WriteLine(converterType.ShortDisplayName());

            PropertyBuilder propEfCore = entityTypeBuilder.Property(property.Name);

            Console.WriteLine($"Property Name: {property.Name}");
            Console.WriteLine($"Property Info Type: {property.PropertyType.Name}");
            Console.WriteLine($"Property CLR Type: {propEfCore.Metadata.ClrType.Name}");
            Console.WriteLine("Declaring Type: {0}", property.DeclaringType.Name);
            Console.WriteLine("Property Name >> {0}", property.Name);

            if (hasNewInitializer)
            {
                if (property.PropertyType.Name == propEfCore.Metadata.ClrType.Name)
                {
                    entityTypeBuilder.Metadata.GetProperty(property.Name)
                        .SetValueConverter(converter);
                }
            }
            else
            {
                entityTypeBuilder.Metadata.GetProperty(property.Name)
                    .SetValueConverter(converter);
            }
        }
    }
}
```

You'll need these helper methods:

```csharp
    public static bool IsVogenValueObject(this Type targetType)
    {
        Maybe<GeneratedCodeAttribute> generatedCodeAttribute = targetType.GetClassAttribute<GeneratedCodeAttribute>();

        return generatedCodeAttribute.HasValue && generatedCodeAttribute.Value.Tool == "Vogen";
    }

    public static Maybe<TAttribute> GetClassAttribute<TAttribute>(this Type targetType)
        where TAttribute : Attribute
    {
        return targetType.GetAttribute<TAttribute>();
    }

    public static Type GetVogenValueObjectType(this Type type)
    {
        if (!type.IsVogenValueObject())
        {
            throw new InvalidDataException("This is not a vogen value object");
        }

        PropertyInfo? valueProperty = type.GetProperty("Value");
        if (valueProperty == null)
        {
            throw new InvalidDataException(
                "We expected to see 'Value' property on the Vogen Value object, but it was not found.  Something is wrong.");
        }

        return valueProperty.PropertyType;
    }

```

Here is the strongly Typed Vogen value object converter which is meant for `Guid`, this is what you might need to re-write a little:

```csharp

public sealed class VogenStronglyTypedIdConverter<TStronglyTypedId, TValue> : ValueConverter<TStronglyTypedId, TValue>
    where TStronglyTypedId : IEquatable<TStronglyTypedId>, IComparable<TStronglyTypedId>
    where TValue : IEquatable<TValue>
{
    public VogenStronglyTypedIdConverter(ConverterMappingHints mappingHints = null)
        : base(valueObject => Serialize(valueObject)!, value => Deserialize(value)!, mappingHints)
    {
    }

    /// <summary>
    ///     Initializes a new instance of the <see cref="VogenStronglyTypedIdConverter{TStronglyTypedId,TValue}" /> type.
    /// </summary>
    public VogenStronglyTypedIdConverter()
        : base(valueObject => Serialize(valueObject)!, value => Deserialize(value)!)
    {
    }

    private static TValue? Serialize(TStronglyTypedId valueObject)
    {
        return valueObject.GetType()
            .GetProperty("Value")
            .GetValue(valueObject, null) is TValue
            ? (TValue)valueObject.GetType()
                .GetProperty("Value")
                .GetValue(valueObject, null)
            : default;

        //TValue value = valueObject.Value;
        //return value;
    }

    private static TStronglyTypedId? Deserialize(TValue value)
    {
        /*if (value.Equals(default))
        {
            return default!;
        }*/

        return (TStronglyTypedId?)(_getFactoryMethod()
            .Invoke(null, new object?[] { value }) ?? throw new InvalidOperationException());

        MethodInfo _getFactoryMethod()
        {
            Type baseSmartEnumObj = _getBaseSmartEnum();

            string methodName = "From";
            MethodInfo? targetMethod = baseSmartEnumObj.GetMethods(BindingFlags.Public | BindingFlags.Static)
                .FirstOrDefault(_ => _.Name == methodName);
            if (targetMethod == null)
            {
                throw new InvalidDataException(
                    $"We could not locate the method '{methodName}' on the type of {baseSmartEnumObj.ShortDisplayName()
                    }");
            }

            return targetMethod;

            Type _getBaseSmartEnum()
            {
                object smartEnum = FormatterServices.GetUninitializedObject(typeof(TStronglyTypedId));

                Type? smartEnumBaseType = smartEnum.GetType();

                if (smartEnumBaseType == null)
                {
                    throw new InvalidDataException("The base Vogen type could not be found");
                }

                return smartEnumBaseType;
            }
        }
    }
}
```


I use it like this within my `IEntityTypeConfiguration` for each entity, but you can also inject it in the `OnModelCreating` method and have success:

```csharp
public abstract class EntityConfiguration<TEntity> : IEntityTypeConfiguration<TEntity>
    where TEntity : Entity<Guid>

{
    ////////////////////////////////////////////////////////////////////////////////////////////////////
    /// <summary>
    ///     Configures the entity of type
    ///     <typeparamref>
    ///         <name>Entity</name>
    ///     </typeparamref>
    /// </summary>
    /// <remarks>   Jeff Ward, 12/6/2021. </remarks>
    /// <param name="entityTypeBuilder">    The builder to be used to configure the entity type. </param>
    /// <seealso cref="M:IEntityTypeConfiguration.Configure(EntityTypeBuilder{TBaseEntity})" />
    ////////////////////////////////////////////////////////////////////////////////////////////////////
    public virtual void Configure(EntityTypeBuilder<TEntity> entityTypeBuilder)
    {
 
       // This is the method
        entityTypeBuilder.UseVogenStronglyTypedId(true);
    }
}

public class EventAccessPasswordConfiguration : EntityConfiguration<EventAccessPassword>
{


    public override void Configure(EntityTypeBuilder<BloomEventAccessPassword> builder)
    {
        base.Configure(builder);
    
       // Your config stuff here....
     
    }
}

```

I spent to long debugging this issue where in some of my classes, I override a `Guid` property primary key with a Vogen Value object, like this:


```csharp
public class WeatherForecast 
{
   public new WeatherForecastId Id { get; set;}
}
```

If you use the `new` keyword, generic configurations will not work.  EfCore gets confused, and will see both your `Guid` and `WeatherForecastId`, then get confused. I have to report the bug to EfCore team tbh.  Even when you declare it like:

```csharp
builder.Entity(entity.ClrType.Name).Property(property.ClrType, propertyInfo.Name).HasConverstion(...);

// or:
builder.Entity(entity.ClrType.Name).Property(property.ClrType, propertyInfo.Name).MetaData.SetValueConverter(...);

```
it still gets confused.

So keep in mind in that case you'll have to manually configure first.

-----

Hope this helped!!

let me know if you have any issues with any of this!


--- 

Sorry my code is not cleaned up with comments and other items, i just finished it a day or two ago.

**Comment by jeffward01** [comment]
You'll also need this if you want to generate values, ill just add it cuz its in the snippets:

```csharp
public class VogenDatabaseValueGenerator<TVogenValueObject> : ValueGenerator<TVogenValueObject>
{
    public override bool GeneratesTemporaryValues => true;

    public override TVogenValueObject Next(EntityEntry entry)
    {
        return (TVogenValueObject?)(_getFactoryMethod()
                                        .Invoke(null, new object?[] { EntityKeyGenerator.GetKey() }) ??
                                    throw new InvalidOperationException(
                                        "The value object could not be generated with the given key.")) ??
               throw new InvalidOperationException("The value object could not be generated with the given key.");

        MethodInfo _getFactoryMethod()
        {
            Type vogenValueObjType = _getVogenValueObject();

            string methodName = "From";
            MethodInfo? targetMethod = vogenValueObjType.GetMethods(BindingFlags.Public | BindingFlags.Static)
                .FirstOrDefault(_ => _.Name == methodName);
            if (targetMethod == null)
            {
                throw new InvalidDataException(
                    $"We could not locate the method '{methodName}' on the type of {vogenValueObjType.ShortDisplayName()
                    }");
            }

            return targetMethod;

            Type _getVogenValueObject()
            {
                object vogenValueObject = FormatterServices.GetUninitializedObject(typeof(TVogenValueObject));

                Type? vogenValueObjectType = vogenValueObject.GetType();

                if (vogenValueObjectType == null)
                {
                    throw new InvalidDataException("The base Vogen type could not be found");
                }

                return vogenValueObjectType;
            }
        }
    }
}
```

UUID v7 - key generators, its nice, but i plan to switch to a `long` one like snowflake soon
and the key generator:
https://github.com/mareek/UUIDNext
```csharp
using UUIDNext;

/// <summary>
///     Generates Id's to be used as primary keys and identifiers
/// </summary>
[PublicAPI]
public static class EntityKeyGenerator
{
    /// <summary>
    ///     Returns a <see cref="Guid" /> created using <see cref="Uuid.NewDatabaseFriendly" />
    /// </summary>
    /// <returns></returns>
    public static Guid GetKey()
    {
        return Uuid.NewDatabaseFriendly();
    }
}
```

**Comment by Blackclaws** [comment]
> > You get queries like this:
> > ```
> >       Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
> >       SELECT COUNT(*)
> >       FROM "Entities" AS "e"
> >       WHERE "e"."ValueObject" = 'Wow'
> > ```
> 
> That SQL looks correct. Im curious why mine had an issue compiling Linq, and yours was fine. Perhaps it was some sort of Expression Compiliation that Ardalis does internally, **before** sending to EfCore, that would do the trick and fix this. That is very interesting and something to explore more. I do not like how I lost my use of `Async` `IQueryable` with the LinqKit solution, and I don't like how I have this line:


You don't have to use Ardalis.Specification at all. That query was generated by this:
```
    var sets = context.Set<TestEntity>().Where(x => x.ValueObject == TestValueObject.From("Wow"));
    Console.WriteLine(sets.Count())
```

The trick is to _not_ compare on .Value at all. Let the ValueConverter handle the conversion into the Db types and compare the value objects directly. Even if you know that the database ultimately stores .Value you _don't_ care about that.

**Comment by Blackclaws** [comment]
@jeffward01 Looking it your code it seems you're configuring the ValueTypes on each entity. What I do is I centrally configure the ValueTypes to have a conversion: 
```

  protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
    {
        base.ConfigureConventions(configurationBuilder);
        AccessControlEfCoreModelConfigurator.ConfigureValueTypes(configurationBuilder);
        configurationBuilder.Properties<ResourceId>().HaveConversion<ResourceId.EfCoreValueConverter>();
        configurationBuilder.Properties<LibraryId>().HaveConversion<LibraryId.EfCoreValueConverter>();
        configurationBuilder.Properties<StorageId>().HaveConversion<StorageId.EfCoreValueConverter>();
        configurationBuilder.Properties<StorageBackendId>().HaveConversion<StorageBackendId.EfCoreValueConverter>();
        configurationBuilder.Properties<LibraryName>().HaveConversion<LibraryName.EfCoreValueConverter>();
        configurationBuilder.Properties<ResourceName>().HaveConversion<ResourceName.EfCoreValueConverter>();
    }
```

That way you can treat any properties as you normally would and the conversion is entirely transparent. What I was looking for was a simple class that combines all the ValueTypes configurations in an assembly and allows me to apply them, preferably source generated :)

 I'm not sure if centrally registering the value converters solves your weather forecast problem where I'm assuming you're actually subclassing a base entity class that you don't control that has a Guid Id. You could try to set up the base property explictly to be ignored when defining the entity configuration. I'm not sure if that's possible though.

**Comment by jeffward01** [comment]
here is a configuration convention you can use, you'll have to modify the code im sure a little. This will save you that repetitive code.
You can even decorate your target ValueObjects with an attribute (for example, I have some VO that I do not use in EfCore), and it will only take the ones decorated.

Also, I throw an exception where `nestedTypes.Length == 0`.  If it is 0, it means there is no EfCore Converter.  


```csharp
internal static class VogenValueObjectConventions
{
    public static void AddVogenEfCoreConverters<TAssemblyMarker>(this ModelConfigurationBuilder configurationBuilder)
        where TAssemblyMarker : class
    {
        Assembly mscorlib = typeof(TAssemblyMarker).Assembly;
        foreach (Type type in mscorlib.GetTypes())
        {
            if (!type.IsVogenValueObject())
            {
                continue;
            }

            Type vogenValueObjectEfCoreConverter = _getVogenValueObjectEfCoreConverter(type);

            configurationBuilder.Properties(type)
                .HaveConversion(vogenValueObjectEfCoreConverter);

            Debug.WriteLine($"(Vogen) | Configured EfCore ValueConverter for ValueObject type >> {type.Name}");
        }

        Type _getVogenValueObjectEfCoreConverter(Type vogenValueObjectType)
        {
            Type[] vogenValueObjectNestedTypes = vogenValueObjectType.GetNestedTypes();

            switch (vogenValueObjectNestedTypes.Length)
            {
                case 1:
                    return vogenValueObjectType.GetNestedTypes()[0];
                case 0:
                    throw new InvalidDataException(
                        $"We expect exactly (1) nested type from the vogen value object, instead we saw (0) nested types. Target Type> {vogenValueObjectType.Name}");
                default:
                    {
                        string listOfTypeNames = string.Join(", ", vogenValueObjectNestedTypes.Select(_ => _.Name));

                        throw new InvalidDataException(
                            $"We expect exactly (1) nested type from the vogen value object, instead we saw ({
                                vogenValueObjectNestedTypes.Length}) nested types. Discovered nested types: {
                                    listOfTypeNames}");
                    }
            }
        }
    }
}
```

And you use it like this:
```csharp
    /// <inheritdoc />
    protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
    {
        configurationBuilder.AddDefaultConventions();
        configurationBuilder.AddVogenEfCoreConverters<IDboAssemblyMarker>();
    }

```

> I'm not sure if centrally registering the value converters solves your weather forecast problem where I'm assuming you're actually subclassing a base entity class that you don't control that has a Guid Id. You could try to set up the base property explictly to be ignored when defining the entity configuration. I'm not sure if that's possible though.

This is correct, its fine for now, I might try that, but I just spent the extra time to configure it by hand... _sigh_ 146 entities ahha.  Go Resharper Live Templates, it was not to bad. But you bring up a good point, and perhaps using private fields will work too.

Another way you can configure it globally using my code i shared previously, is to just run it like this:

```csharp
    protected override void OnModelCreating(ModelBuilder builder)
    {
        Console.WriteLine("Init OnModelCreating");
        Debug.WriteLine("Init OnModelCreating");
        base.OnModelCreating(builder);
        this.ConfigureIdentityEntities(builder);
        builder.ApplyEntityConfiguration();
        this.ApplyEfConfigs(builder);
        this.SetModelFilters(builder);



        foreach (IMutableEntityType entityType in builder.Model.GetEntityTypes()
                     .ToList())
        {

            builder.Entity(entityType.ClrType).UseVogenStronglyTypedId();
            
            // If you have issues with the above one ^^ try this one below:
            builder.Entity(entityType.Name).UseVogenStronglyTypedId();
        }
}
```

The configuration thing i sent first will also work

Let me know if you have any issues with these!! Good luck! I appreciate the conversation and passing of ideas

**Comment by SteveDunn** [comment]
Hi - I'm trying to close the remaining issues labelled as bug. Is this still an issue, or do the workarounds above satisfy the original issue reported?

I understand the friction involved with using types that Vogen generate for the domain layer and then being used in the infrastructure layer, but I also understand the usefulness of being able to use them everywhere.

Maybe we can somehow utilise interceptors in C# 12. But for now now, I'd like to close this one if there are no objections.

**Comment by SteveDunn** [comment]
Closed as I don't _think_ this is a really a bug in Vogen, but please raise the alarm if I am completely wrong!
